feat: GitHub-hosted runners for benchmarks

HowardBraham opened the pull request:
## **Description**

### Major changes
- **benchmarks.yml** - Does some important and new-to-us things like:
  - ~~Self-hosted runner `runs-on: gha-mm-scale-set-ubuntu-22.04-amd64-med`~~
  - ~~Running in the same kind of Docker container we use in CircleCI `container: image: cimg/node:22.13-browsers`~~
  - ~~Unlike on CircleCI, does not work without installing `Xvfb`~~
- **codespaces.yml**
  - I originally added this workflow, and it was important to caching in the past, but with the new workflows in GHA, it's no longer needed, so I deleted it
- **test-circular-deps.yml**
  - We were hitting the workflow limit of 20, so I inlined this simple workflow, that did not need to be reusable


### Prerequisites to merging this PR
- ✔️ Merge this `benchmarks` branch of `github-tools`: https://github.com/MetaMask/github-tools/blob/benchmarks/.github/actions/setup-environment/action.yml
- ✔️ Change `setup-environment@benchmarks` to `setup-environment@1d657e262aea7e3f216754febb624831527d2565`
- ~~Vastly increase the number of allocated self-hosted runners~~ (no longer needed because not using self-hosted anymore)

### This is just Part 1 of a larger 4-part task to make the startup time quality gate, but I think it's the hardest part
1. Run on GitHub-hosted runners so that hopefully the numbers will be more stable (this PR)
2. Get this to measure the same measurements in the same way as we do for Sentry
3. Make metamaskbot post with the results of this instead of the current CircleCI source
4. Set up the quality gate

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/29955?quickstart=1)

## **Related issues**

Progresses: MetaMask/MetaMask-planning#3679

<!--
## **Manual testing steps**
## **Screenshots/Recordings**
### **Before**
### **After**
## **Pre-merge author checklist**
## **Pre-merge reviewer checklist**
-->

HowardBraham commented on .github/workflows/benchmarks.yml:
```diff
@@ -0,0 +1,42 @@
+        with:
+          name: ${{ inputs.name }}
+          path: test-artifacts/chrome/benchmark/
+          retention-days: 5
```
The number of `retention-days` was chosen very arbitrarily, and maybe someone has an opinion

HowardBraham replied:
These particular benchmark artifacts are very very small.  It's just some JSON data that gets zipped.  On one run, it's 823 bytes and 196 bytes.  But we also may want to look back at these numbers much later.  GitHub probably isn't the best platform for that though, something like Sentry or Segment is probably much better.

HowardBraham commented on .github/workflows/update-attributions.yml:
```diff
@@ -182,4 +182,4 @@ jobs:
           GITHUB_TOKEN: ${{ secrets.LAVAMOAT_UPDATE_TOKEN }}
           PR_NUMBER: ${{ github.event.issue.number }}
-          ACTION_RUN_URL: "${{ github.server_url }}/${{ github.repository }}/actions/runs/${{ github.run_id }}"
+          ACTION_RUN_URL: '${{ github.server_url }}/${{ github.repository }}/actions/runs/${{ github.run_id }}'
```
This is just Prettier running on a file

Gudahtt commented:
Could you split this into multiple PRs? A lot of these changes seem to be distinct from each other

HowardBraham committed:
ci: self-hosted runners for benchmarks

HowardBraham commented:
As requested by @itsyoboieltr, here are benchmarks running in 4 different configurations:
- browserify build on self-hosted runners
- browserify build on ubuntu-latest
- webpack build on self-hosted runners
- webpack build on ubuntu-latest

Probably the most important stat here is **uiStartup** (btw our old system highlights **load**).  Two additional things I see:
- Standard deviation is pretty low across all configurations, which is nice
- All configurations are approximately twice as fast as our CircleCI machine

I'm going to run this again to see if it varies over time, and also with some other options.

<table><thead><tr><th>Platform</th><th>Page</th><th>Metric</th><th>Min (ms)</th><th>Max (ms)</th><th>Average (ms)</th><th>StandardDeviation (ms)</th><th>MarginOfError (ms)</th></tr></thead><tbody><tr><td rowspan="11">Browserify-self-host</td><td rowspan="11">Home</td><td>firstPaint</td><td align="right">158</td><td align="right">1195</td><td align="right">714</td><td align="right">302</td><td align="right">145</td></tr><tr><td>domContentLoaded</td><td align="right">821</td><td align="right">1177</td><td align="right">868</td><td align="right">75</td><td align="right">36</td></tr><tr><td>load</td><td align="right">824</td><td align="right">1179</td><td align="right">873</td><td align="right">74</td><td align="right">36</td></tr><tr><td>domInteractive</td><td align="right">14</td><td align="right">27</td><td align="right">20</td><td align="right">2</td><td align="right">1</td></tr><tr><td>backgroundConnect</td><td align="right">5</td><td align="right">11</td><td align="right">7</td><td align="right">1</td><td align="right">1</td></tr><tr><td>firstReactRender</td><td align="right">12</td><td align="right">17</td><td align="right">12</td><td align="right">1</td><td align="right">1</td></tr><tr><td>getState</td><td align="right">3</td><td align="right">5</td><td align="right">4</td><td align="right">1</td><td align="right">0</td></tr><tr><td>initialActions</td><td align="right">0</td><td align="right">1</td><td align="right">0</td><td align="right">0</td><td align="right">0</td></tr><tr><td>loadScripts</td><td align="right">618</td><td align="right">809</td><td align="right">655</td><td align="right">42</td><td align="right">20</td></tr><tr><td>setupStore</td><td align="right">6</td><td align="right">9</td><td align="right">7</td><td align="right">1</td><td align="right">0</td></tr><tr><td>uiStartup</td><td align="right">899</td><td align="right">1292</td><td align="right">971</td><td align="right">81</td><td align="right">39</td></tr><tr><td rowspan="11">Browserify-ubuntu-latest</td><td rowspan="11">Home</td><td>firstPaint</td><td align="right">216</td><td align="right">1061</td><td align="right">869</td><td align="right">299</td><td align="right">143</td></tr><tr><td>domContentLoaded</td><td align="right">965</td><td align="right">1144</td><td align="right">1016</td><td align="right">40</td><td align="right">19</td></tr><tr><td>load</td><td align="right">973</td><td align="right">1150</td><td align="right">1022</td><td align="right">40</td><td align="right">19</td></tr><tr><td>domInteractive</td><td align="right">26</td><td align="right">69</td><td align="right">36</td><td align="right">14</td><td align="right">7</td></tr><tr><td>backgroundConnect</td><td align="right">8</td><td align="right">12</td><td align="right">10</td><td align="right">1</td><td align="right">1</td></tr><tr><td>firstReactRender</td><td align="right">18</td><td align="right">37</td><td align="right">21</td><td align="right">4</td><td align="right">2</td></tr><tr><td>getState</td><td align="right">5</td><td align="right">11</td><td align="right">8</td><td align="right">2</td><td align="right">1</td></tr><tr><td>initialActions</td><td align="right">0</td><td align="right">1</td><td align="right">0</td><td align="right">0</td><td align="right">0</td></tr><tr><td>loadScripts</td><td align="right">751</td><td align="right">874</td><td align="right">791</td><td align="right">30</td><td align="right">15</td></tr><tr><td>setupStore</td><td align="right">9</td><td align="right">14</td><td align="right">11</td><td align="right">1</td><td align="right">1</td></tr><tr><td>uiStartup</td><td align="right">1090</td><td align="right">1371</td><td align="right">1169</td><td align="right">63</td><td align="right">30</td></tr><tr><td rowspan="11">Webpack-self-host</td><td rowspan="11">Home</td><td>firstPaint</td><td align="right">154</td><td align="right">736</td><td align="right">668</td><td align="right">119</td><td align="right">57</td></tr><tr><td>domContentLoaded</td><td align="right">666</td><td align="right">718</td><td align="right">685</td><td align="right">15</td><td align="right">7</td></tr><tr><td>load</td><td align="right">670</td><td align="right">734</td><td align="right">692</td><td align="right">15</td><td align="right">7</td></tr><tr><td>domInteractive</td><td align="right">22</td><td align="right">44</td><td align="right">25</td><td align="right">5</td><td align="right">2</td></tr><tr><td>backgroundConnect</td><td align="right">16</td><td align="right">24</td><td align="right">19</td><td align="right">2</td><td align="right">1</td></tr><tr><td>firstReactRender</td><td align="right">11</td><td align="right">13</td><td align="right">12</td><td align="right">0</td><td align="right">0</td></tr><tr><td>getState</td><td align="right">3</td><td align="right">6</td><td align="right">5</td><td align="right">1</td><td align="right">0</td></tr><tr><td>initialActions</td><td align="right">0</td><td align="right">1</td><td align="right">0</td><td align="right">0</td><td align="right">0</td></tr><tr><td>loadScripts</td><td align="right">656</td><td align="right">710</td><td align="right">676</td><td align="right">15</td><td align="right">7</td></tr><tr><td>setupStore</td><td align="right">4</td><td align="right">7</td><td align="right">5</td><td align="right">1</td><td align="right">0</td></tr><tr><td>uiStartup</td><td align="right">804</td><td align="right">882</td><td align="right">833</td><td align="right">17</td><td align="right">8</td></tr><tr><td rowspan="11">Webpack-ubuntu-latest</td><td rowspan="11">Home</td><td>firstPaint</td><td align="right">170</td><td align="right">896</td><td align="right">520</td><td align="right">318</td><td align="right">153</td></tr><tr><td>domContentLoaded</td><td align="right">708</td><td align="right">890</td><td align="right">829</td><td align="right">35</td><td align="right">17</td></tr><tr><td>load</td><td align="right">714</td><td align="right">898</td><td align="right">836</td><td align="right">36</td><td align="right">17</td></tr><tr><td>domInteractive</td><td align="right">21</td><td align="right">40</td><td align="right">34</td><td 
align="right">4</td><td align="right">2</td></tr><tr><td>backgroundConnect</td><td align="right">22</td><td align="right">29</td><td align="right">25</td><td align="right">2</td><td align="right">1</td></tr><tr><td>firstReactRender</td><td align="right">13</td><td align="right">15</td><td align="right">13</td><td align="right">1</td><td align="right">0</td></tr><tr><td>getState</td><td align="right">3</td><td align="right">8</td><td align="right">6</td><td align="right">1</td><td align="right">1</td></tr><tr><td>initialActions</td><td align="right">0</td><td align="right">0</td><td align="right">0</td><td align="right">0</td><td align="right">0</td></tr><tr><td>loadScripts</td><td align="right">697</td><td align="right">879</td><td align="right">818</td><td align="right">35</td><td align="right">17</td></tr><tr><td>setupStore</td><td align="right">5</td><td align="right">8</td><td align="right">6</td><td align="right">1</td><td align="right">0</td></tr><tr><td>uiStartup</td><td align="right">883</td><td align="right">1049</td><td align="right">998</td><td align="right">33</td><td align="right">16</td></tr></tbody></table>

HowardBraham commented:
Second run of this, numbers are pretty close to the first run

<table><thead><tr><th>Platform</th><th>Page</th><th>Metric</th><th>Min (ms)</th><th>Max (ms)</th><th>Average (ms)</th><th>StandardDeviation (ms)</th><th>MarginOfError (ms)</th></tr></thead><tbody><tr><td rowspan="11">Browserify-self-host</td><td rowspan="11">Home</td><td>firstPaint</td><td align="right">157</td><td align="right">905</td><td align="right">776</td><td align="right">200</td><td align="right">96</td></tr><tr><td>domContentLoaded</td><td align="right">802</td><td align="right">868</td><td align="right">833</td><td align="right">16</td><td align="right">8</td></tr><tr><td>load</td><td align="right">807</td><td align="right">875</td><td align="right">839</td><td align="right">17</td><td align="right">8</td></tr><tr><td>domInteractive</td><td align="right">14</td><td align="right">24</td><td align="right">19</td><td align="right">2</td><td align="right">1</td></tr><tr><td>backgroundConnect</td><td align="right">5</td><td align="right">8</td><td align="right">7</td><td align="right">1</td><td align="right">0</td></tr><tr><td>firstReactRender</td><td align="right">11</td><td align="right">13</td><td align="right">12</td><td align="right">1</td><td align="right">0</td></tr><tr><td>getState</td><td align="right">3</td><td align="right">5</td><td align="right">4</td><td align="right">1</td><td align="right">0</td></tr><tr><td>initialActions</td><td align="right">0</td><td align="right">1</td><td align="right">0</td><td align="right">0</td><td align="right">0</td></tr><tr><td>loadScripts</td><td align="right">602</td><td align="right">669</td><td align="right">630</td><td align="right">17</td><td align="right">8</td></tr><tr><td>setupStore</td><td align="right">5</td><td align="right">7</td><td align="right">6</td><td align="right">1</td><td align="right">0</td></tr><tr><td>uiStartup</td><td align="right">893</td><td align="right">986</td><td align="right">938</td><td align="right">23</td><td align="right">11</td></tr><tr><td rowspan="11">Browserify-ubuntu-latest</td><td rowspan="11">Home</td><td>firstPaint</td><td align="right">209</td><td align="right">1107</td><td align="right">863</td><td align="right">322</td><td align="right">155</td></tr><tr><td>domContentLoaded</td><td align="right">960</td><td align="right">1082</td><td align="right">1008</td><td align="right">34</td><td align="right">16</td></tr><tr><td>load</td><td align="right">963</td><td align="right">1091</td><td align="right">1014</td><td align="right">35</td><td align="right">17</td></tr><tr><td>domInteractive</td><td align="right">17</td><td align="right">64</td><td align="right">36</td><td align="right">15</td><td align="right">7</td></tr><tr><td>backgroundConnect</td><td align="right">7</td><td align="right">16</td><td align="right">10</td><td align="right">2</td><td align="right">1</td></tr><tr><td>firstReactRender</td><td align="right">18</td><td align="right">37</td><td align="right">22</td><td align="right">4</td><td align="right">2</td></tr><tr><td>getState</td><td align="right">5</td><td align="right">12</td><td align="right">7</td><td align="right">2</td><td align="right">1</td></tr><tr><td>initialActions</td><td align="right">0</td><td align="right">1</td><td align="right">0</td><td align="right">0</td><td align="right">0</td></tr><tr><td>loadScripts</td><td align="right">754</td><td align="right">856</td><td align="right">788</td><td align="right">29</td><td align="right">14</td></tr><tr><td>setupStore</td><td align="right">9</td><td align="right">23</td><td align="right">11</td><td align="right">3</td><td align="right">1</td></tr><tr><td>uiStartup</td><td align="right">1107</td><td align="right">1265</td><td align="right">1152</td><td align="right">40</td><td align="right">19</td></tr><tr><td rowspan="11">Webpack-self-host</td><td rowspan="11">Home</td><td>firstPaint</td><td align="right">172</td><td align="right">752</td><td align="right">677</td><td align="right">117</td><td align="right">56</td></tr><tr><td>domContentLoaded</td><td align="right">655</td><td align="right">745</td><td align="right">695</td><td align="right">21</td><td align="right">10</td></tr><tr><td>load</td><td align="right">669</td><td align="right">754</td><td align="right">705</td><td align="right">20</td><td align="right">9</td></tr><tr><td>domInteractive</td><td align="right">22</td><td align="right">70</td><td align="right">28</td><td align="right">10</td><td align="right">5</td></tr><tr><td>backgroundConnect</td><td align="right">17</td><td align="right">24</td><td align="right">19</td><td align="right">2</td><td align="right">1</td></tr><tr><td>firstReactRender</td><td align="right">11</td><td align="right">13</td><td align="right">12</td><td align="right">1</td><td align="right">0</td></tr><tr><td>getState</td><td align="right">3</td><td align="right">6</td><td align="right">4</td><td align="right">1</td><td align="right">0</td></tr><tr><td>initialActions</td><td align="right">0</td><td align="right">1</td><td align="right">0</td><td align="right">0</td><td align="right">0</td></tr><tr><td>loadScripts</td><td align="right">647</td><td align="right">737</td><td align="right">687</td><td align="right">21</td><td align="right">10</td></tr><tr><td>setupStore</td><td align="right">4</td><td align="right">7</td><td align="right">5</td><td align="right">1</td><td align="right">1</td></tr><tr><td>uiStartup</td><td align="right">822</td><td align="right">897</td><td align="right">847</td><td align="right">20</td><td align="right">10</td></tr><tr><td rowspan="11">Webpack-ubuntu-latest</td><td rowspan="11">Home</td><td>firstPaint</td><td align="right">169</td><td align="right">878</td><td align="right">458</td><td align="right">307</td><td align="right">148</td></tr><tr><td>domContentLoaded</td><td align="right">805</td><td align="right">895</td><td align="right">830</td><td align="right">21</td><td align="right">10</td></tr><tr><td>load</td><td align="right">812</td><td align="right">904</td><td align="right">837</td><td align="right">21</td><td align="right">10</td></tr><tr><td>domInteractive</td><td align="right">26</td><td align="right">81</td><td align="right">37</td><td align="right">12</td><td align="right">6</td></tr><tr><td>backgroundConnect</td><td align="right">21</td><td align="right">27</td><td align="right">24</td><td align="right">2</td><td align="right">1</td></tr><tr><td>firstReactRender</td><td align="right">12</td><td align="right">14</td><td align="right">13</td><td align="right">0</td><td align="right">0</td></tr><tr><td>getState</td><td align="right">4</td><td align="right">10</td><td align="right">6</td><td align="right">1</td><td align="right">1</td></tr><tr><td>initialActions</td><td align="right">0</td><td align="right">1</td><td align="right">0</td><td align="right">0</td><td align="right">0</td></tr><tr><td>loadScripts</td><td align="right">794</td><td align="right">883</td><td align="right">819</td><td align="right">21</td><td align="right">10</td></tr><tr><td>setupStore</td><td align="right">5</td><td align="right">8</td><td align="right">6</td><td align="right">1</td><td align="right">0</td></tr><tr><td>uiStartup</td><td align="right">963</td><td align="right">1081</td><td align="right">991</td><td align="right">26</td><td align="right">13</td></tr></tbody></table>

HowardBraham commented on .github/workflows/benchmarks.yml:
```diff
@@ -0,0 +1,43 @@
+      - name: Setup environment
+        uses: metamask/github-tools/.github/actions/setup-environment@benchmarks
+
+      - name: Download artifact prep-build-test-webpack
+        uses: actions/download-artifact@v4
+        with:
+          path: ./dist/
+          pattern: prep-build-test-webpack
+          merge-multiple: true
```
That would be cool, but I think GitHub Actions doesn't allow this.  If you find documentation that says otherwise, let's go for it.

Another option would be a naming scheme like perhaps `main-benchmarks.yml`

HowardBraham commented on test/e2e/webdriver/chrome.js:
```diff
@@ -61,6 +61,11 @@ class ChromeDriver {
 
+    // It will crash if you don't do this, but there might be another way around it
+    if (process.env.GITHUB_ACTION) {
+      args.push('--no-sandbox');
```
I ran it again for you, here's the full run log: https://github.com/MetaMask/metamask-extension/actions/runs/13192830006/job/36828840527

Important snippet:
```
[763:763:0207/035354.645008:FATAL:zygote_host_impl_linux.cc(126)] No usable sandbox! Update your kernel or see https://chromium.googlesource.com/chromium/src/+/main/docs/linux/suid_sandbox_development.md for more information on developing with the SUID sandbox. If you want to live dangerously and need an immediate workaround, you can try using --no-sandbox.
[0207/035354.652173:ERROR:file_io_posix.cc(145)] open /sys/devices/system/cpu/cpu0/cpufreq/scaling_cur_freq: No such file or directory (2)
[0207/035354.652233:ERROR:file_io_posix.cc(145)] open /sys/devices/system/cpu/cpu0/cpufreq/scaling_max_freq: No such file or directory (2)
/__w/metamask-extension/metamask-extension/node_modules/selenium-webdriver/lib/error.js:524
    let err = new ctor(data.message)
              ^
SessionNotCreatedError: session not created: Chrome failed to start: exited normally.
  (session not created: DevToolsActivePort file doesn't exist)
  (The process started from chrome location /github/home/.cache/selenium/chrome/linux64/1[26](https://github.com/MetaMask/metamask-extension/actions/runs/13192830006/job/36828840527#step:7:27).0.6478.182/chrome is no longer running, so ChromeDriver is assuming that Chrome has crashed.)
```

HowardBraham replied:
@NicholasEllul @itsyoboieltr this issue for puppeteer (which we're not using here, but is similar to Selenium) gave me a good clue:
https://github.com/puppeteer/puppeteer/issues/12818#issuecomment-2415915338

I tried it with our setup, and indeed `runs-on: ubuntu-22.04` works.

Is there any objection to us using `runs-on: ubuntu-22.04` until someone else fixes this?

HowardBraham committed:
responding to reviews

HowardBraham commented on .github/workflows/main.yml:
```diff
@@ -16,7 +17,7 @@ on:
   lint-workflows:
     name: Lint workflows
-    uses: metamask/github-tools/.github/workflows/lint-workflows.yml@c534f265e02af2f2422a3c686bb09a11bfbf4cc2
+    uses: metamask/github-tools/.github/workflows/lint-workflows.yml@main
```
Yeah it should be a new hash.  Should I make all `@main` into `@hash` right now?

HowardBraham replied:
Done

HowardBraham commented on .github/workflows/benchmark-matrix.yml:
```diff
@@ -0,0 +1,48 @@
+    [
+      'yarn benchmark:chrome --out test-artifacts/chrome/benchmark/pageload.json --retries 2',
+      'yarn user-actions-benchmark:chrome --out test-artifacts/chrome/benchmark/user_actions.json --retries 2',
+    ]
```
Haha thanks, that took a lot of R&D to get right!

HowardBraham commented on .github/workflows/benchmark-matrix.yml:
```diff
@@ -0,0 +1,48 @@
+  benchmarks:
+    runs-on: ubuntu-latest
+    container:
+      image: cimg/node:22.13-browsers
```
`ubuntu-latest` is GitHub-hosted
`gha-mm-scale-set-ubuntu-22.04-amd64-med` is self-hosted

HowardBraham committed:
needs: run-benchmark-matrix

HowardBraham committed:
ubuntu-22.04 and headless=new

HowardBraham committed:
replaced all setup-environment@main and lint-workflows.yml@main with @hash

HowardBraham commented on .github/workflows/benchmark-matrix.yml:
```diff
@@ -0,0 +1,42 @@
+        with:
+          name: benchmark-${{ matrix.buildType }}-${{ matrix.testType }}
+          path: test-artifacts/chrome/benchmark/
+          retention-days: 30
```
Just looked it up, the default is 90 days.  Do we want the benchmark results to stay around for 90 days?  They will also get posted by metamaskbot and so sorta stay around forever.

HowardBraham commented on .github/workflows/main.yml:
```diff
@@ -66,10 +76,52 @@ jobs:
+        with:
+          name: prep-build-test-browserify
+          path: ./dist/
+          retention-days: 1
```
Yes, `yarn download-builds:test`, which I wrote, does exactly this from CircleCI.  I think when we migrate all builds off of CircleCI, we want to increase this, but for now I think 1 day is fine.

HowardBraham committed:
env.COMMANDS as JSON
removed "Will enable this as a required job in a later PR"
multiline list
new github-tools hash

HowardBraham committed:
default retention-days

HowardBraham committed:
Apply suggestions from itsyoboieltr code review

Co-authored-by: Norbert Elter <72046715+itsyoboieltr@users.noreply.github.com>

HowardBraham committed:
rename to run-benchmarks.yml

itsyoboieltr reviewed (approved):
LGTM! Great work!

NEllusion reviewed (approved):
LGTM! Regarding the 90 day retention period, we can monitor to see the costs and what savings we may get by switching to a lower number later on.
